File: Node.js/VK_Cup_2016_Qualification_Round_2/road_improvement.js
/*
 * node --version -> v4.2.1
 */

var readline = require('readline');

var counter = amount = 0;
var accumulation = [];

function accumulator(road_number, road) {
  if (accumulation.length == 0) {
    accumulation[0] = [[road_number], road];
    return;
  }

  for (var idx in accumulation) {
    var item = accumulation[idx];
    if (item[1].indexOf(road[0]) == -1 && item[1].indexOf(road[1]) == -1) {
      item[0].push(road_number);
      item[1] = item[1].concat(road);
      return;
    }
  }

  accumulation.push([[road_number], road]);
}

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

rl.on('line', function(line) {
  if (counter == 0) {
    amount = parseInt(line);
  } else {
    accumulator(counter, line.trim().split(' '));
  }

  counter++;
  
  if (counter >= amount) {
    rl.close();

    console.log(accumulation.length);
    for (var idx in accumulation) {
      var item = accumulation[idx];
      console.log(item[0].length + ' ' + item[0].join(' '));
    }
  }
});

File: algorithm/www.codeforces.com/Python_2.7.10/VK_Cup_2016_Qualification_Round_2/road_improvement.py
#
## python --version -> Python 2.7.10
#

import sys

amount = 0

accumulation = []

def accumulator(road_number, road):
  if len(accumulation) == 0:
    accumulation.append([[road_number], road])
    return

  for item in accumulation:
    if road[0] not in item[1] and road[1] not in item[1]:
      item[0].append(road_number)
      item[1] = item[1] + road
      return

  accumulation.append([[road_number], road])


amount = int(raw_input())

tmp = 1
while tmp < amount:
  accumulator(tmp, raw_input().strip().split(' '))
  tmp += 1



print len(accumulation)

for item in accumulation:
  sys.stdout.write(str(len(item[0])) + ' ' + ' '.join(str(e) for e in item[0]) + '\n')


